### PR TITLE
[REVIEW] Split out cudf::distinct_count from drop_duplicates.cu

### PR DESCRIPTION
**Pull request by davidwendt**
One of our longest compile times is for `drop_duplicates.cu` measured currently at about 15 minutes. This file contains two cudf APIs: `cudf::drop_duplicates` and `cudf::distinct_count`. They share no code so it is reasonable to move the `distinct_count` to its own source file. Both files individually have a compile time around 7 minutes.

Also, analyzing the `drop_duplicates` source code I found that the table row-comparator is used in two `thrust::copy_if()` calls which inlines this large comparator function many times. And, the first `copy_if` has a lambda with two calls to the comparator meaning it is inlined twiced for each time it is inlined by `copy_if`. This is the main cause of the increased compile time and code size. I found a way to combine the logic for both the `copy_if` lambdas and further reduce the lambda to include only one copy of the comparator. The `copy_if` itself will still inline this many times but should reduce it by a factor of 3.

The `drop_duplicates` does not have a gbenchmark so I added one in this PR to measure before and after this change. No measurable effect was found so I'm including the change in this PR.

The `distinct_count` API passes the row-comparator to the `thrust::count_if` but only once so the same improvement does not apply there.

Note that no new function has been added (or removed). This is simply an improvement to compile time and size for this source file. The compile time for `drop_duplicates.cu` is now less than 5 minutes.

### CODE REVIEW

File: cpp/src/stream_compaction/distinct_count.cu [hunk 1]
```diff
@@ -0,0 +1,196 @@
+/*
```
Comment (davidwendt) [inline cpp/src/stream_compaction/distinct_count.cu:1]: This is not new code but was just moved from the `drop_duplicates.cu` source file here to its own source file.

File: cpp/src/stream_compaction/drop_duplicates.cu [hunk 2]
```diff
@@ -151,44 +173,6 @@ column_view get_unique_ordered_indices(cudf::table_view const& keys,
   }
 }
 
-cudf::size_type distinct_count(table_view const& keys,
```
Comment (davidwendt) [inline cpp/src/stream_compaction/drop_duplicates.cu:154]: This code was moved to the new `distinct_count.cu` source file.

**Comment by kkraus14** [comment]
@davidwendt did you do any experimentation with removing pragmas from thrust to see what happens if things aren't inlined in this situation and the performance implications?

**Comment by davidwendt** [comment]
> @davidwendt did you do any experimentation with removing pragmas from thrust to see what happens if things aren't inlined in this situation and the performance implications?

Unroll may have implications for [`thrust::copy_if`](https://github.com/NVIDIA/thrust/blob/main/thrust/system/cuda/detail/copy_if.h#L341-L355) (https://godbolt.org/z/asffbo) which is used by code in this PR but I've not tried disabling it here.

**Comment by davidwendt** [comment]
rerun tests

**Review by kkraus14 — APPROVED** [review]
cmake lgtm

**Review by harrism — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

**Pull request merged**